perf(api): Omit feature flags from API views of multiple projects

RyanSkonnord opened the pull request:
Suppress the list of feature flags from the "all projects" API endpoint,
and from the projects nested in a team. This is intended to improve
performance where an organization has many projects.

This may be considered a breaking change, so for now the change occurs
only behind a feature flag for experimental performance improvements.

RyanSkonnord committed:
perf(api): Omit feature flags from API views of multiple projects

Suppress the list of feature flags from the "all projects" API endpoint,
and from the projects nested in a team. This is intended to improve
performance where an organization has many projects.

This may be considered a breaking change, so for now the change occurs
only behind a feature flag for experimental performance improvements.

EvanPurkhiser commented on src/sentry/api/serializers/models/project.py:
```diff
@@ -165,6 +166,9 @@ def get_feature_list(self, obj, user):
         from sentry.features.base import ProjectFeature
 
+        if not self.include_features:
+            return None
```
I wonder if it would be good to also kill the key in the response `serialize`

RyanSkonnord replied:
I thought it was doing that actually. Fixed in f274062

leedongwei reviewed (commented):
Looks alright to this frontend dev, tagging `app-backend` for confirmation.

RyanSkonnord committed:
fixup! perf(api): Omit feature flags from API views of multiple projects

Have "features" attribute left out of the JSON response entirely, rather
than being a null value.

dcramer reviewed (commented):
If its unused in the frontend no need to worry about compatibility. Feature flags are considered internal, so we can break it.

RyanSkonnord committed:
fixup! perf(api): Omit feature flags from API views of multiple projects

wedamija reviewed (commented):
Instead of a feature flag, would it make sense to have the frontend pass a variable through that causes this to not get sent? That way you can easily test this on an enterprise org without risking breaking anything for their users.

Also wondering if these are actually unused - I'd have assumed we'd use these to decide which features to show on the frontend, but I'm not familiar with the frontend at all so 🤷

RyanSkonnord commented:
@wedamija 

> Instead of a feature flag, would it make sense to have the frontend pass a variable through that causes this to not get sent? That way you can easily test this on an enterprise org without risking breaking anything for their users.

The feature flag is meant to be temporary, for internal testing. We'd eventually get rid of the feature flag and make this the universal behavior.

However, controlling it with a parameter like you say is under consideration. The way I see it, it comes down to an API design choice. Which is preferable: to add the complexity of an extra parameter to the request interface, or to add the complexity of a conditional attribute to the response interface? I'm very much open to opinions here.

> Also wondering if these are actually unused - I'd have assumed we'd use these to decide which features to show on the frontend, but I'm not familiar with the frontend at all so 🤷

I believe they are used only when querying for a single project. The hope is that the client code never cares about which features are enabled on individual projects when querying for a big batch of them. (The internal testing will make sure.)

markstory commented:
> I believe they are used only when querying for a single project. The hope is that the client code never cares about which features are enabled on individual projects when querying for a big batch of them. (The internal testing will make sure.)

The list of projects is fed into the reflux ProjectsStore which we use in a variety of places across the frontend to get cached access to individual projects.  There are several UI components (withProject, withProjects, utils/projects) that read from reflux and feed other components. If we aren't using `project.features` anywhere we should be safe.

If we aren't using project.features in the UI should we even continue to support project level feature flags? I don't remember a time in recent history where we elected to use a project level feature instead of an organization level one.

RyanSkonnord commented:
> The list of projects is fed into the reflux ProjectsStore which we use in a variety of places across the frontend to get cached access to individual projects. There are several UI components (withProject, withProjects, utils/projects) that read from reflux and feed other components.

Thanks @markstory, that's important background. I'd like us (@getsentry/enterprise) to do some follow-up research on that. If it turns out there is a need to populate the reflux ProjectsStore with project flags, that could be solved with a request parameter to lighten the response as discussed above.

leedongwei commented:
Thanks for the context. I did a quick audit of the project-level feature flags: `issue-alerts-targeting` was recently used in the last month and the rest have been dormant for 1+ year.

The PR doesn't have breaking changes to the API as it is. We should merge this and measure performance improvements and see if it's worth the investment to rethink/refactor project-level feature flags across the frontend/backend.

leedongwei reviewed (approved):
😎

wedamija commented:
> @wedamija
> 
> > Instead of a feature flag, would it make sense to have the frontend pass a variable through that causes this to not get sent? That way you can easily test this on an enterprise org without risking breaking anything for their users.
> 
> The feature flag is meant to be temporary, for internal testing. We'd eventually get rid of the feature flag and make this the universal behavior.
> 
> However, controlling it with a parameter like you say is under consideration. The way I see it, it comes down to an API design choice. Which is preferable: to add the complexity of an extra parameter to the request interface, or to add the complexity of a conditional attribute to the response interface? I'm very much open to opinions here.

I think it's more about what's useful - would you like to be able to test this on say, riot or square without turning the feature flag on for their org (and so not changing things for their users)? Or are you happy just enabling this flag on any project you want to test on? I don't have an opinion, just wanted to bring it up.
